### PR TITLE
API: change datetimelike Index to raise IndexError instead ValueError

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
I noticed this in the plotting PR that DatetimeIndex et al behave a bit strange on an invalid index:

```
In [41]: pd.Index([1, 2, 3])['no_int']
...
IndexError: only integers, slices (`:`), ellipsis (`...`), numpy.newaxis (`None`) and integer or boolean arrays are valid indices

In [42]: pd.DatetimeIndex([1, 2, 3])['no_int']
...
ValueError: 
```

I think this should be consistent IndexError ?

### CODE REVIEW

File: pandas/tests/indexes/test_base.py [hunk 1]
```diff
@@ -591,12 +591,19 @@ def test_empty_fancy(self):
...
-        arr = np.array(self.dateIndex)
-        exp = self.dateIndex[5]
-        exp = _to_m8(exp)
+    def test_getitem(self, indices):
 
-        assert exp == arr[5]
+        if indices.size != 100:
+            return
+
+        exp = getattr(indices, '_box_func', lambda x: x)(indices._values[0])
```
Comment (jorisvandenbossche) [inline pandas/tests/indexes/test_base.py]: This is a bit too much 'internal-details-relying' for a test, but I didn't directly know a way to write it generic for all types of index instead (or maybe we just shouldn't try, and keep it only testing the errors)

File: pandas/tests/indexes/test_base.py [hunk 2]
```diff
@@ -591,12 +591,19 @@ def test_empty_fancy(self):
...
             pytest.raises(IndexError, idx.__getitem__, empty_farr)
 
-    def test_getitem(self):
-        arr = np.array(self.dateIndex)
-        exp = self.dateIndex[5]
-        exp = _to_m8(exp)
+    def test_getitem(self, indices):
 
-        assert exp == arr[5]
+        if indices.size != 100:
```
Comment (jreback) [inline pandas/tests/indexes/test_base.py]: why do you have this?

File: pandas/tests/indexes/test_base.py (hunk 1 above)
Comment (jreback) [reply pandas/tests/indexes/test_base.py]: you don't need this at all, just check for the error

File: pandas/tests/indexes/test_base.py [hunk 3]
```diff
@@ -591,12 +591,19 @@ def test_empty_fancy(self):
             # Index.
             pytest.raises(IndexError, idx.__getitem__, empty_farr)
 
-    def test_getitem(self):
```
Comment (jreback) [inline pandas/tests/indexes/test_base.py]: this original test is pretty useless, so no need to keep its function (and this is tested else in many places as well)
all of the indexes have a ``test_indexing.py``

Comment (jreback) [reply pandas/tests/indexes/test_base.py]: just rename the test to ``test_getitem_error``. if you want you can split this out into a new file, ``test_indexing.py`` in this dir.

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — CHANGES_REQUESTED** [review]
add a note in whatsnew indexing bug fixes (0.22)

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @jorisvandenbossche

**Pull request merged**